pkg/cmd/pipeline/list.go

```diff
@@ -16,51 +16,171 @@ package pipeline
 
 import (
 	"fmt"
+	"io"
 	"os"
+	"text/tabwriter"
 
 	"github.com/spf13/cobra"
 	"github.com/tektoncd/cli/pkg/cli"
+	"github.com/tektoncd/cli/pkg/formatted"
+	"github.com/tektoncd/cli/pkg/printer"
 	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
-	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	cliopts "k8s.io/cli-runtime/pkg/genericclioptions"
 )
 
-func ListCommand(p cli.Params) *cobra.Command {
+const (
+	emptyMsg = "No pipelines found"
+	header   = "NAME\tAGE\tLAST RUN\tSTARTED\tDURATION\tSTATUS"
+	body     = "%s\t%s\t%s\t%s\t%s\t%s\n"
+	blank    = "---"
+)
 
-	defaultOutput := `jsonpath={range .items[*]}{.metadata.name}{"\n"}{end}`
-	f := cliopts.NewPrintFlags("").WithDefaultOutput(defaultOutput)
+func listCommand(p cli.Params) *cobra.Command {
+	f := cliopts.NewPrintFlags("list")
 
 	c := &cobra.Command{
 		Use:     "list",
 		Aliases: []string{"ls"},
 		Short:   "Lists pipelines in a namespace",
 		Long:    ``,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			ps, err := list(p)
-
+			output, err := cmd.LocalFlags().GetString("output")
 			if err != nil {
-				fmt.Fprintf(os.Stderr, "Failed to list pipelines from %s namespace\n", p.Namespace())
+				fmt.Fprint(os.Stderr, "Error: output option not set properly \n")
 				return err
 			}
 
-			printer, err := f.ToPrinter()
-			if err != nil {
-				return err
+			if output != "" {
+				return printPipelineListObj(cmd.OutOrStdout(), p, f)
 			}
-			return printer.PrintObj(ps, cmd.OutOrStdout())
+
+			return printPipelineDetails(cmd.OutOrStdout(), p)
 		},
 	}
 	f.AddFlags(c)
 
 	return c
 }
 
-func list(p cli.Params) (*v1alpha1.PipelineList, error) {
+func printPipelineDetails(out io.Writer, p cli.Params) error {
+
+	ps, prs, err := listPipelineDetails(p)
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to list pipelines from %s namespace\n", p.Namespace())
+		return err
+	}
+
+	if len(ps.Items) == 0 {
+		fmt.Fprintln(out, emptyMsg)
+		return nil
+	}
+
+	w := tabwriter.NewWriter(out, 0, 5, 3, ' ', tabwriter.TabIndent)
+	fmt.Fprintln(w, header)
+
+	for _, pipeline := range ps.Items {
+		pr, ok := prs[pipeline.Name]
+		if !ok {
+			fmt.Fprintf(w, body,
+				pipeline.Name,
+				formatted.Age(pipeline.CreationTimestamp, p.Time()),
+				blank,
+				blank,
+				blank,
+				blank,
+			)
+			continue
+		}
+
+		fmt.Fprintf(w, body,
+			pipeline.Name,
+			formatted.Age(pipeline.CreationTimestamp, p.Time()),
+			pr.Name,
+			formatted.Age(*pr.Status.StartTime, p.Time()),
+			formatted.Duration(pr.Status.StartTime, pr.Status.CompletionTime),
+			formatted.Condition(pr.Status.Conditions[0]),
+		)
+
+	}
+	return w.Flush()
+}
+
+func printPipelineListObj(w io.Writer, p cli.Params, f *cliopts.PrintFlags) error {
+	ps, err := listAllPipelines(p)
+
+	if err != nil {
+		return err
+	}
+	return printer.PrintObject(w, ps, f)
+}
+
+func listAllPipelines(p cli.Params) (*v1alpha1.PipelineList, error) {
```

Comment:
nit: for the future, those "get data from client" shouldn't need the `cli.Params` I feel (here we just need the clients and the namespace)

Reply:
Do you have reason `why` it is better to do it that way i.e pass in `client, ns` opposed to `cli.Params` ? 
Also, I don't have a strong opinion why `p cli.Param` should be passed, so I will make that change in this PR itself.

Reply:
As of today no *important* reason **but** we may/will want to list pipelines in multiple case. This function might get extracted elsewhere *and* in that case, it shouldn't be tied to any `cli` code.
I tend to like having the `cli` related code constrained to only a really small subset of packages (aka, in our case `pkg/cmd`).

But it's a nit, no worries for that PR :wink:

Reply:
I have implemented the change anyway to see how decoupling feels like.

Reply:
I am marking it as resolved. Please reopen if you disagree